sale_order_merge/models/sale_order_merge.py

```diff
@@ -0,0 +1,198 @@
+        for invoice in (
+                self.sale_order.invoice_ids +
+                self.to_merge.mapped('invoice_ids')):
+            if invoice.state == 'draft' and not invoice.internal_number:
+                if target:
+                    other_inv += invoice
+                else:
+                    target = invoice
+            else:
+                keep_inv += invoice
+        if target:
+            other_inv.mapped('invoice_line').write({'invoice_id': target.id})
+            other_inv.mapped('tax_line').write({'invoice_id': target.id})
+            other_inv.unlink()
+            target.button_compute(set_total=True)
+
+        for inv in target + keep_inv:
+            self.sale_order.write({'invoice_ids': [(4, inv.id)]})
+        self.to_merge.write({'invoice_ids': [(6, 0, [])]})
+
+    @api.multi
+    def _picking_can_merge(self, picking):
+        return (picking.state not in ('done', 'cancel') and
+                picking.location_dest_id.usage == 'customer')
+
+    @api.multi
+    def _get_picking_map_key(self, picking):
+        return (picking.picking_type_id, picking.location_id,
+                picking.location_dest_id, picking.partner_id)
+
+    @api.multi
+    def merge_pickings(self):
+        """ Assign all pickings to the target sale order and merge any
+        pending pickings """
+        orders = self.sale_order + self.to_merge
+        group = self.env['procurement.group']
+        if self.sale_order.procurement_group_id:
+            group = self.sale_order.procurement_group_id
+        else:
+            for order in self.to_merge:
+                if order.procurement_group_id:
+                    group = order.procurement_group_id
+                    break
+            else:
+                return  # no group, no pickings
+            self.sale_order.write({'procurement_group_id': group.id})
+        other_groups = orders.mapped('procurement_group_id')
+        self.env['stock.picking'].search(
+            [('group_id', 'in', other_groups.ids)]).write(
+                {'group_id': group.id})
+        self.env['stock.move'].search(
+            [('group_id', 'in', other_groups.ids)]).write(
+                {'group_id': group.id})
+        self.env['procurement.order'].search(
+            [('group_id', 'in', other_groups.ids)]).write(
+                {'group_id': group.id})
+        pick_map = {}
+        for picking in self.sale_order.picking_ids:
+            if self._picking_can_merge(picking):
+                key = self._get_picking_map_key(picking)
+                if key not in pick_map:
+                    pick_map[key] = self.env['stock.picking']
+                pick_map[key] += picking
+        for pickings in pick_map.values():
+            if len(pickings) > 1:
+                target = pickings[0]
+                pickings -= target
+                pickings.mapped('move_lines').write({'picking_id': target.id})
+                pickings.unlink()
+        return True
```

Comment:
The final picking should have as Source Document the final sales order, not the original.

![image](https://cloud.githubusercontent.com/assets/7683926/18946325/a980bee8-862e-11e6-99c1-4b75aebfe95e.png)